File: pe-core/src/main/java/org/saar/core/renderer/uniforms/UniformsHelper.java
package org.saar.core.renderer.uniforms;

import org.saar.lwjgl.opengl.shader.uniforms.UniformWrapper;

import java.util.ArrayList;
import java.util.List;

public abstract class UniformsHelper {

    public static UniformsHelper empty() {
        return Empty.EMPTY;
    }

    public abstract UniformsHelper addUniform(UniformWrapper uniform);

    public abstract void load();

    private static class Empty extends UniformsHelper {

        private static final Empty EMPTY = new Empty();

        @Override
        public UniformsHelper addUniform(UniformWrapper uniform) {
            final UniformsHelper helper = new Generic();
            return helper.addUniform(uniform);
        }

        @Override
        public void load() {
        }
    }

    private static class Generic extends UniformsHelper {

        private final List<UniformWrapper> uniforms = new ArrayList<>();

        @Override
        public UniformsHelper addUniform(UniformWrapper uniform) {
            this.uniforms.add(uniform);
            return this;
        }

        @Override
        public void load() {
            this.uniforms.forEach(UniformWrapper::load);
        }
    }

}

File: pe-core/src/main/java/org/saar/core/renderer/uniforms/UniformsHelper.kt
package org.saar.core.renderer.uniforms

import org.saar.lwjgl.opengl.shader.uniforms.UniformWrapper

interface UniformsHelper {
    fun addUniform(uniform: UniformWrapper): UniformsHelper

    fun load()

    private object Empty : UniformsHelper {
        override fun addUniform(uniform: UniformWrapper) = Generic().addUniform(uniform)

        override fun load() = Unit
    }

    private class Generic : UniformsHelper {
        private val uniforms = mutableListOf<UniformWrapper>()

        override fun addUniform(uniform: UniformWrapper) = this.also { this.uniforms.add(uniform) }

        override fun load() = this.uniforms.forEach(UniformWrapper::load)
    }

    companion object {
        fun empty(): UniformsHelper = Empty
    }
}
